docs/madr/decisions/003-logging.md

```diff
@@ -0,0 +1,220 @@
+# Logging standard
+
+* Status: accepted
+
+## Context and Problem Statement
+
+Seamless troubleshooting is a feature. We should keep more attention to the logging in Kuma.
+
+## Considered Options
+
+* Adopting consistent logging standard.
+
+## Decision Outcome
+
+Chosen option: "Adopting consistent logging standard", because it's the only option.
+
+### Positive Consequences
+
+* Better UX for users.
+* Better troubleshooting means both users and maintainers spend less time debugging issues.
+
+### Negative Consequences
+
+* More effort to try to follow the guidelines.
+
+## Pros and Cons of the Options
+
+### Adopting consistent logging standard
+
+#### Convention
+
+Ideally, all those points are linted.
+
+1) Logs are lowercase. Can be uppercase when you start with proper noun.
+_GOOD_
+```go
+log.Info("something has happened")
+log.Info("Envoy is starting")
+```
+_BAD_
+```go
+log.Info("Something has happened")
+```
+
+2) Do not include `.` at the end of the sentence if this is the only sentence in the log
+_GOOD_
+```go
+log.Info("something has happened")
+log.Info("something has happened. Restarting the process")
+```
+_BAD_
+```go
+log.Info("something has happened.")
+```
+
+3) Either log or return an error. Don't do both, unless you want to log more details on DEBUG level.
+_GOOD_
+```go
+if err != nil {
+	return errors.Wrap(err, "more info")
+}
+...
+if err != nil {
+	log.Error(err, "more info")
+	return nil
+}
+...
+if err != nil {
+	log.V(1).Info("much more details about the error")
```

Comment:
Should the level be `DEBUG` as mentioned on line [56](https://github.com/kumahq/kuma/pull/4445/files#diff-03543e676073151a7aba162f34cbd764dc512a896e8f7a5bf2880bea0fe0bb85R56)?

Reply:
Ok, I've just read [this](https://github.com/kumahq/kuma/pull/4445/files#diff-03543e676073151a7aba162f34cbd764dc512a896e8f7a5bf2880bea0fe0bb85R134) and it's so confusing. Why do we have `V(1).Info` which actually means `Debug`?

Reply:
Yes, `log.V(1).Info` means logging on `debug` level. `log.V(0).Info` means logging on `info` level. `V(0)` is explicit.

Reply:
For me it's confusing to read something like

```
log...Info(...)
```

and find out that it actually means "debug". I think we either base our "levels" on numbers (0, 1...) or on words (info, debug ...)

Reply:
Maybe we can wrap the `go-logr` interface and restrict it to `Info` == `V(0).Info`, `Debug` == `V(1).Info`.

As for conditionally doing `V(1).WithValues` maybe we can add a `WithDebugValues` which checks if `V(1)` before calling `WithValues`? It'd at least cover the common case where the value is already evaluated.

Reply:
Added clarification on how to use loggers on different levels.

For the wrapper. There is a slight chance that we will need something custom to implement this https://github.com/kumahq/kuma/issues/4446

I'm slightly anxious that the wrapper would affect the performance
https://github.com/go-logr/logr/blob/master/logr.go#L231

Reply:
@michaelbeaumont if you feel strongly about this logger wrapper, please file a separate issue since it's a bit out of scope of this PR (improvement of API, not the standard itself).